Replace `require_login` with Pundit in Webui::Users::NotificationsCon…

intrip opened the pull request:
This is a PR of a series which replaces `require_login` with `Pundit`.
You can find further relevant info in #10083.

Tackles `Webui::Users::NotificationsController`

Ref #10083

<!---
If you haven't done so already, please read the CONTRIBUTING.md file to learn
how we work and what we expect from all contributors.

https://github.com/openSUSE/open-build-service/blob/master/CONTRIBUTING.md

In order to make it as easy as possible for other developers to review your
pull request we ask you to:

- Explain what this PR is about in the description
- Explain the steps the reviewer has to follow to verify your change
- If the reviewer needs sample data to verify your change, please explain how to
  create that data
- If you include visual changes in this PR, please add screenshots or GIFs
- If you address performance in this PR, add benchmark data or explain how the
  reviewer can benchmark this

This is a good PR description example:

Hey Friends,

this introduces labels for the different build result states on the project
monitor page. This makes it easier to get a visual overview of what is going on
in your project.

To verify this feature

- Enable the interconnect to build.opensuse.org
- Create the project home:Admin
- Add 'openSUSE Tumbleweed' as a repository to the project
- Branch a couple of packages into the project:
  ```
  for i in `osc -A http://0.0.0.0:3000 ls openSUSE.org:home:hennevogel`; do osc -A http://0.0.0.0:3000 copypac openSUSE.org:home:hennevogel $i home:Admin; done
  ```
- Visit the monitor page and see the new labels for the different states.

Here is a screenshot of how it looks:

** Before **
![Screenshot of the project monitor](https://example.com/screenshot1.png)

** After **
![Screenshot of the project monitor](https://example.com/screenshot2.png)

-->

intrip commented on src/api/app/policies/notification_policy.rb:
```diff
@@ -1,4 +1,16 @@
+    true
+  end
+
+  def update?
+    true
+  end
```
I'm not totally convinced of this approach thus I'd like to further discuss it: the idea is that since we introduced `verify_policy_scoped` in the controller checking for the `subscriber` in `NotificationPolicy#update` it's redundant and also leads to a slowdown of the related endpoint (trade-off), moreover we also have tests in the controller to ensure we cannot update other user notifications. On the other hand if we look at the `NotificationPolicy#update` per-se we don't see that constraint and therefore is misleading.

intrip commented:
@dmarcoux could you take a look here? I've followed an approach to authorize collections but I'd like to know your take on this.

dmarcoux commented:
We are already [scoping](https://github.com/varvet/pundit#scopes) the notifications to the currently logged-in user in the `fetch_notifications` method https://github.com/openSUSE/open-build-service/blob/6e1e25f6aed699c00bc9fcf8d1ff0090bc62ba70/src/api/app/controllers/webui/users/notifications_controller.rb#L75-L82

This is how we handled authorization for collections of `Notification` in `NotificationPolicy`. `fetch_notifications` is then called in the `index` and `update` actions, through `paginated_notifications` or directly.

Now, to get rid of `require_login`, I wonder if only setting up the `initialize` method as you did is enough. I believe it would be, but I'm not sure how it behaves for anonymous users. This would have to be tested. If `initialize` is enough, the methods `index?` and `update?` in `NotificationPolicy` won't be needed since we're scoping the notifications as I described above.

intrip commented:
@dmarcoux 
Agree with you about the scoping , I've also added `verify_policy_scoped` and [a test](https://github.com/intrip/open-build-service/blob/refactor_require_login_user_notifications/src/api/spec/controllers/webui/users/notifications_controller_spec.rb#L132-L143) to make sure we won't introduce regressions on that side (the `index` was already covered).

Related to `require_login` I think that the `initialize` check is enough: I've added both [index tests](https://github.com/intrip/open-build-service/blob/refactor_require_login_user_notifications/src/api/spec/controllers/webui/users/notifications_controller_spec.rb#L30-L33) and [update tests](https://github.com/intrip/open-build-service/blob/refactor_require_login_user_notifications/src/api/spec/controllers/webui/users/notifications_controller_spec.rb#L110-L113) to ensure that everything works as expected for anonymous users.
My main concerns were on the readability of the `NotificationPolicy`: having 2 methods (`update?`, `index?`) that always returns `true` is misleading.
I took a deeper look at `Pundit` implementation, since `authorize` [requires (or infers) a `query`](https://github.com/varvet/pundit/blob/v2.1.0/lib/pundit.rb#L216) I went for a little different approach: we just [initialize the policy in a filter](https://github.com/intrip/open-build-service/blob/refactor_require_login_user_notifications/src/api/app/controllers/webui/users/notifications_controller.rb#L7), this allow us to get rid of both the `index?` and `update?` methods in the policy.

Could you please review the changes?

dmarcoux commented:
Using a closed scope is a great idea.
